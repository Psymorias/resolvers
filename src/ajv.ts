import { appendErrors, transformToNestObject, Resolver } from 'react-hook-form';
import Ajv from 'ajv';


const parseErrorSchema = (
  error: Array<Ajv.ErrorObject>,
  validateAllFieldCriteria: boolean,
) =>
  error.reduce(
    (previous: Record<string, any>, { dataPath, message = '', keyword }) => ({
      ...previous,
      ...(dataPath
        ? previous[dataPath] && validateAllFieldCriteria
          ? {
            [dataPath]: appendErrors(
              dataPath,
              validateAllFieldCriteria,
              previous,
              keyword,
              message,
            ),
          }
          : {
            [dataPath]: previous[dataPath] || {
              message,
              type: keyword,
              ...(validateAllFieldCriteria
                ? {
                  types: { [keyword]: message || true },
                }
                : {}),
            },
          }
        : {}),
    }),
    {},
  );

export const ajvResolver = <TFieldValues extends Record<string, any>>(
  schema: object,
  options: Ajv.Options = {
    allErrors: true
  },
): Resolver<TFieldValues> => async (
  values,
  _,
  validateAllFieldCriteria = false,
) => {
  const ajv = new Ajv({
    ...options
  });

  const validate = ajv.compile({
    '$async': true,
    ...schema
  });

  try {
    return {
      values: (await validate(values)) as any,
      errors: {},
    };
  } catch (e) {
    return {
      values: {},
      errors: transformToNestObject(
        parseErrorSchema(e.errors, validateAllFieldCriteria),
      ),
    };
  }
};
