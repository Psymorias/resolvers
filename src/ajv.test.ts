import { ajvResolver } from './ajv';

const errors = {
  message: "validation failed",
  errors: [
    {
      keyword: "minLength",
      dataPath: ".firstName",
      schemaPath: "#/properties/firstName/minLength",
      params: {
        limit: 1
      },
      message: "should NOT be shorter than 1 characters"
    },
    {
      keyword: "minLength",
      dataPath: ".email",
      schemaPath: "#/properties/email/minLength",
      params: {
        limit: 1
      },
      message: "should NOT be shorter than 1 characters"
    },
    {
      keyword: "format",
      dataPath: ".email",
      schemaPath: "#/properties/email/format",
      params: {
        format: "email"
      },
      message: "should match format \"email\""
    },
    {
      keyword: "enum",
      dataPath:".options",
      schemaPath: "#/properties/options/enum",
      params: {
        allowedValues: ["Option 1", "Option 2"]
      },
      message:"should be equal to one of the allowed values"
    }
  ],
  validation: true,
  ajv: true
};

const schema = {
  type: "object",
  properties: {
    firstName: { type: "string", minLength: 1 },
    email: { type: "string", format: "email", minLength: 1, maxLength: 50 },
    options: { type: "string", enum: ["Option 1", "Option 2"] }
  },
  required: [
    "firstName",
    "email",
    "options"
  ],
  additionalProperties: false
}

describe('ajvResolver', () => {
  it('should get values', async () => {
    const data = {
      firstName: 'jimmy',
      email: 'test@test.de',
      options: 'Option 1'
    };
    expect(await ajvResolver(schema)(data)).toEqual({
      errors: {},
      values: {
        firstName: 'jimmy',
        email: 'test@test.de',
        options: 'Option 1'
      },
    });
  });

  it('should get errors', async () => {
    const data = {
      firstName: 2,
      email: undefined,
      options: 'three'
    };
    expect(await ajvResolver(schema)(data)).toMatchSnapshot();
  });
});

describe('ajv - validateWithSchema', () => {
  it('should return undefined when no error reported', async () => {
    expect(
      await ajvResolver({
        validate: () => {
          throw errors;
        },
      } as any)({}),
    ).toMatchSnapshot();
  });

  it('should return empty object when validate pass', async () => {
    expect(
      await ajvResolver({
        validate: () => new Promise((resolve) => resolve()),
      } as any)({}),
    ).toEqual({
      errors: {},
      values: {},
    });
  });
});
