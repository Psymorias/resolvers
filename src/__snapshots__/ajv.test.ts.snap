// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ajv - validateWithSchema should return undefined when no error reported 1`] = `
Object {
  "errors": Object {},
  "values": Object {},
}
`;

exports[`ajvResolver should get errors 1`] = `
Object {
  "errors": Object {
    "email": Object {
      "message": "should be string",
      "type": "type",
    },
    "firstName": Object {
      "message": "should have required property 'firstName'",
      "type": "required",
    },
    "options": Object {
      "message": "should be equal to one of the allowed values",
      "type": "enum",
    },
  },
  "values": Object {},
}
`;

exports[`ajvResolver should localize errors 1`] = `
Object {
  "errors": Object {
    "email": Object {
      "message": "muss das erforderliche Attribut email enthalten",
      "type": "required",
    },
    "firstName": Object {
      "message": "muss sein: string",
      "type": "type",
    },
    "options": Object {
      "message": "muss einem der vorgegebenen Werte entsprechen",
      "type": "enum",
    },
  },
  "values": Object {},
}
`;
